File: scripts/parity-probe.mjs
// TypeScript-path parity probe: runs the SAME deterministic tensor through
// best.onnx via onnxruntime-web (wasm, in Node) and decodes with the same
// math as src/lib/model/postprocess.ts. Compare output to ml/parity_python.json.
//
// Run:  node scripts/parity-probe.mjs   ->  ml/parity_web.json

import { readFileSync, writeFileSync } from "node:fs";
import * as ort from "onnxruntime-web";

const SIZE = 640;
const CONF = 0.25;
const IOU = 0.45;
const ANCHORS = 8400;

function lcgImage() {
  const n = 3 * SIZE * SIZE;
  const out = new Float32Array(n);
  let state = 123456789n;
  const A = 1103515245n, C = 12345n, M = 0x7fffffffn;
  for (let i = 0; i < n; i++) {
    state = (A * state + C) & M;
    out[i] = Number(state % 256n) / 255;
  }
  return out;
}

function iou(a, b) {
  const ix1 = Math.max(a[0], b[0]), iy1 = Math.max(a[1], b[1]);
  const ix2 = Math.min(a[2], b[2]), iy2 = Math.min(a[3], b[3]);
  const inter = Math.max(0, ix2 - ix1) * Math.max(0, iy2 - iy1);
  if (!inter) return 0;
  const ua =
    (a[2] - a[0]) * (a[3] - a[1]) + (b[2] - b[0]) * (b[3] - b[1]) - inter;
  return inter / ua;
}

function decode(out) {
  const cands = [];
  for (let i = 0; i < ANCHORS; i++) {
    let best = 0, cls = -1;
    for (let c = 0; c < 3; c++) {
      const s = out[(4 + c) * ANCHORS + i];
      if (s > best) { best = s; cls = c; }
    }
    if (best < CONF) continue;
    const cx = out[i], cy = out[ANCHORS + i], w = out[2 * ANCHORS + i], h = out[3 * ANCHORS + i];
    cands.push({ cls, conf: best, box: [cx - w / 2, cy - h / 2, cx + w / 2, cy + h / 2] });
  }
  const kept = [];
  for (let c = 0; c < 3; c++) {
    const clsC = cands.filter((d) => d.cls === c).sort((a, b) => b.conf - a.conf);
    const sup = new Array(clsC.length).fill(false);
    for (let i = 0; i < clsC.length; i++) {
      if (sup[i]) continue;
      kept.push(clsC[i]);
      for (let j = i + 1; j < clsC.length; j++) {
        if (!sup[j] && iou(clsC[i].box, clsC[j].box) > IOU) sup[j] = true;
      }
    }
  }
  kept.sort((a, b) => b.conf - a.conf);
  return kept.map((d) => ({
    cls: d.cls,
    conf: +d.conf.toFixed(5),
    box: d.box.map((v) => +v.toFixed(2)),
  }));
}

const buf = readFileSync("public/models/best.onnx");
const session = await ort.InferenceSession.create(buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength), {
  executionProviders: ["wasm"],
});
const data = lcgImage();
const input = new ort.Tensor("float32", data, [1, 3, SIZE, SIZE]);
const { output0 } = await session.run({ images: input });
const y = output0.data;

let sum = 0;
for (let i = 0; i < y.length; i++) sum += y[i];
let boxSum = 0;
for (let i = 0; i < 4 * ANCHORS; i++) boxSum += y[i];
let scoreMax = 0, scoreSum = 0;
for (let i = 4 * ANCHORS; i < 7 * ANCHORS; i++) {
  scoreSum += y[i];
  if (y[i] > scoreMax) scoreMax = y[i];
}
let inputSum = 0;
for (let i = 0; i < data.length; i++) inputSum += data[i];

const result = {
  inputChecksum: +inputSum.toFixed(2),
  outputShape: [1, 7, ANCHORS],
  outputStats: {
    sum: +sum.toFixed(1),
    boxMean: +(boxSum / (4 * ANCHORS)).toFixed(4),
    scoreMax: +scoreMax.toFixed(5),
    scoreMean: +(scoreSum / (3 * ANCHORS)).toFixed(6),
  },
  detections: decode(y),
};
writeFileSync("ml/parity_web.json", JSON.stringify(result, null, 2));
console.log(JSON.stringify(result, null, 2));

File: ml/parity_probe.py
"""Numeric parity probe: Python vs TypeScript decode path.

Feeds a deterministic synthetic 640x640 tensor (identical bytes to the JS
probe, same LCG) through best.onnx and dumps raw output stats + decoded boxes
using the same math pipeline.py relies on. scripts/parity-probe.mjs does the
identical thing via onnxruntime-web; compare the two JSON files.

This isolates the real parity risks — tensor layout, channel order, decode
strides, NMS — from image-resampling differences (cv2 vs canvas), which are
bounded separately by the on-footage check.

Run:  uv run python parity_probe.py   ->  parity_python.json
"""

import json

import numpy as np
import onnxruntime as ort

SIZE = 640
CONF = 0.25
IOU = 0.45


def lcg_image() -> np.ndarray:
    """Deterministic RGB float tensor via a 32-bit LCG (same as JS probe)."""
    n = 3 * SIZE * SIZE
    out = np.empty(n, dtype=np.float32)
    state = 123456789
    for i in range(n):
        state = (1103515245 * state + 12345) & 0x7FFFFFFF
        out[i] = (state % 256) / 255.0
    return out.reshape(1, 3, SIZE, SIZE)


def decode(output: np.ndarray):
    """Same decode as src/lib/model/postprocess.ts (identity meta: no letterbox)."""
    out = output[0]  # [7, 8400]
    boxes, scores, classes = [], [], []
    cls_scores = out[4:7]
    best = cls_scores.max(axis=0)
    best_cls = cls_scores.argmax(axis=0)
    keep = best >= CONF
    for i in np.flatnonzero(keep):
        cx, cy, w, h = out[0, i], out[1, i], out[2, i], out[3, i]
        boxes.append([float(cx - w / 2), float(cy - h / 2), float(cx + w / 2), float(cy + h / 2)])
        scores.append(float(best[i]))
        classes.append(int(best_cls[i]))

    # greedy per-class NMS
    kept = []
    for c in set(classes):
        idxs = [i for i in range(len(boxes)) if classes[i] == c]
        idxs.sort(key=lambda i: -scores[i])
        while idxs:
            i = idxs.pop(0)
            kept.append(i)
            idxs = [j for j in idxs if iou(boxes[i], boxes[j]) <= IOU]
    kept.sort(key=lambda i: -scores[i])
    return [
        {
            "cls": classes[i],
            "conf": round(scores[i], 5),
            "box": [round(v, 2) for v in boxes[i]],
        }
        for i in kept
    ]


def iou(a, b):
    ix1, iy1 = max(a[0], b[0]), max(a[1], b[1])
    ix2, iy2 = min(a[2], b[2]), min(a[3], b[3])
    inter = max(0, ix2 - ix1) * max(0, iy2 - iy1)
    if inter == 0:
        return 0.0
    ua = (a[2] - a[0]) * (a[3] - a[1]) + (b[2] - b[0]) * (b[3] - b[1]) - inter
    return inter / ua


def main():
    sess = ort.InferenceSession("best.onnx", providers=["CPUExecutionProvider"])
    x = lcg_image()
    y = sess.run(None, {"images": x})[0]

    result = {
        "inputChecksum": round(float(x.sum()), 2),
        "outputShape": list(y.shape),
        "outputStats": {
            "sum": round(float(y.sum()), 1),
            "boxMean": round(float(y[0, :4].mean()), 4),
            "scoreMax": round(float(y[0, 4:7].max()), 5),
            "scoreMean": round(float(y[0, 4:7].mean()), 6),
        },
        "detections": decode(y),
    }
    with open("parity_python.json", "w") as f:
        json.dump(result, f, indent=2)
    print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()
